File: backend/routers/terminal.py
"""
Terminal API — full trading terminal: live quotes, order book, order placement.
"""

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel
from loguru import logger

from backend.broker.kite_auth import kite_session
from backend.broker.orders import order_manager

router = APIRouter(prefix="/terminal", tags=["terminal"])


class OrderRequest(BaseModel):
    symbol: str
    transaction_type: str    # BUY or SELL
    quantity: int
    price: float = 0.0       # 0 = market order
    trigger_price: float = 0.0
    product: str = "MIS"
    order_type: str = "LIMIT"


class GTTRequest(BaseModel):
    symbol: str
    direction: str
    quantity: int
    stop_loss: float
    target: float
    ltp: float


@router.get("/quote/{symbol}")
async def get_live_quote(symbol: str):
    """Real-time quote with OHLC, volume, depth."""
    try:
        quote = kite_session.get_quote([f"NSE:{symbol}"])
        data = quote.get(f"NSE:{symbol}", {})
        depth = kite_session.get_market_depth(symbol)
        return {
            "symbol": symbol,
            "ltp": data.get("last_price"),
            "ohlc": data.get("ohlc", {}),
            "volume": data.get("volume"),
            "change": data.get("net_change"),
            "depth": depth,
        }
    except Exception as e:
        raise HTTPException(status_code=502, detail=str(e))


@router.post("/order")
async def place_order(body: OrderRequest):
    """Place a manual order from the terminal."""
    try:
        if body.transaction_type.upper() == "BUY":
            if body.price > 0:
                order_id = order_manager.place_buy_order(body.symbol, body.quantity, body.price)
            else:
                order_id = order_manager.place_market_order(body.symbol, "BUY", body.quantity)
        else:
            if body.price > 0:
                order_id = order_manager.place_sell_order(body.symbol, body.quantity, body.price)
            else:
                order_id = order_manager.place_market_order(body.symbol, "SELL", body.quantity)
        return {"status": "ORDER_PLACED", "order_id": order_id}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/gtt")
async def set_gtt(body: GTTRequest):
    """Set a GTT order (stop loss + target pair)."""
    try:
        gtt_id = order_manager.set_gtt(
            symbol=body.symbol,
            direction=body.direction,
            quantity=body.quantity,
            stop_loss=body.stop_loss,
            target=body.target,
            ltp=body.ltp,
        )
        return {"status": "GTT_SET", "gtt_id": gtt_id}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/orders")
async def get_orders():
    """All orders placed today."""
    try:
        orders = kite_session.kite.orders()
        return {"orders": orders}
    except Exception as e:
        raise HTTPException(status_code=502, detail=str(e))


@router.delete("/order/{order_id}")
async def cancel_order(order_id: str):
    """Cancel a pending order."""
    try:
        order_manager.cancel_order(order_id)
        return {"status": "CANCELLED", "order_id": order_id}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/margins")
async def get_margins():
    """Available margin/funds."""
    try:
        margins = kite_session.kite.margins()
        equity = margins.get("equity", {})
        return {
            "available": equity.get("available", {}).get("live_balance", 0),
            "used": equity.get("utilised", {}).get("debits", 0),
            "net": equity.get("net", 0),
        }
    except Exception as e:
        raise HTTPException(status_code=502, detail=str(e))

File: frontend/src/api/terminal.ts
import { api } from './client'

export interface LiveQuote {
  symbol: string
  ltp: number
  ohlc: { open: number; high: number; low: number; close: number }
  volume: number
  change: number
  depth: {
    buy:  { price: number; quantity: number; orders: number }[]
    sell: { price: number; quantity: number; orders: number }[]
  }
}

export interface OrderRequest {
  symbol: string
  transaction_type: 'BUY' | 'SELL'
  quantity: number
  price?: number
  trigger_price?: number
  product?: 'MIS' | 'CNC' | 'NRML'
  order_type?: 'LIMIT' | 'MARKET' | 'SL' | 'SL-M'
}

export interface GTTRequest {
  symbol: string
  direction: 'LONG' | 'SHORT'
  quantity: number
  stop_loss: number
  target: number
  ltp: number
}

export interface Margins {
  available: number
  used: number
  net: number
}

export const terminalApi = {
  getQuote:   (symbol: string) => api.get<LiveQuote>(`/terminal/quote/${symbol}`).then((r) => r.data),
  placeOrder: (body: OrderRequest) => api.post('/terminal/order', body).then((r) => r.data),
  setGTT:     (body: GTTRequest)   => api.post('/terminal/gtt', body).then((r) => r.data),
  getOrders:  () => api.get('/terminal/orders').then((r) => r.data),
  cancelOrder:(id: string) => api.delete(`/terminal/order/${id}`).then((r) => r.data),
  getMargins: () => api.get<Margins>('/terminal/margins').then((r) => r.data),
  getCandles: (symbol: string, interval: string, days = 5) =>
    api.get(`/stocks/${symbol}/candles`, { params: { interval, days } }).then((r) => r.data),
}
